Src/Support/Google.Apis.Auth/OAuth2/ServiceAccountCredential.cs

```diff
@@ -365,7 +386,7 @@ private Task<string> GetOrCreateJwtAccessTokenAsync(string authUri)
         /// <param name="issueUtc">The issue time of the JWT.</param>
         /// <param name="expiryUtc">The expiry time of the JWT.</param>
         /// </summary>
-        private string CreateJwtAccessToken(string authUri, DateTime issueUtc, DateTime expiryUtc)
+        private string CreateJwtAccessTokenWithAudience(string authUri, DateTime issueUtc, DateTime expiryUtc)
```

Comment:
I suggest this whole method to be as follows and the we can get rid of the CreateJwtAccessTokenWithScopes method.

```
private string CreateJwtAccessToken(string authUri, DateTime issueUtc, DateTime expiryUtc)
{
    var payload = new GoogleJsonWebSignature.Payload()
    {
        Issuer = Id,
        Subject = Id,
        IssuedAtTimeSeconds = (long)(issueUtc - UnixEpoch).TotalSeconds,
        ExpirationTimeSeconds = (long)(expiryUtc - UnixEpoch).TotalSeconds,
    };
    if (HasExplicitScopes)
    {
        payload.Scope = Scopes.Aggregate((a, b) => a + " " + b);
    }
    else
    {
        payload.Audience = authUri;
    }
    return CreateAssertionFromPayload(payload);
}
```

Reply:
That would work for me, yes. (Would string.Join be simpler than Scopes.Aggregate though?)

Reply:
The proposed method is valid and cuts some lines for sure. However, I think my original proposal is easier to understand, which, I think, could work better for a public library. Let me describe my line of thought, and we can drop it if it is not convincing enough :) 

The benefit of the original implementation is that all the Jwt logic that checks Scopes, UseJwtWithScopes flag and decides on payload is within couple methods that call each other: GetAccessTokenForRequestAsync -> GetOrCreateJwtAccessTokenAsync. Its hard to miss something out.

The suggested solution moves some logic into CreateJwtAccessToken and as a result:
 (a) overall chain is longer: GetAccessTokenForRequestAsync -> GetOrCreateJwtAccessTokenAsync -> CreateJwtAccessToken
(b) the middle method does not provide a hint that you have to look into the CreateJwtAccessToken to complete the puzzle :) 

Let me know if this makes sense :)

Reply:
I would prefer keeping the cache logic independent of what is being cached (see my comment about passing the calculated cache key also).

In my suggestion you have one method that decides whether to generate a JWT or an auth token, that is `GetAccessTokenForRequestAsync` and one method that knows how to generate a JWT, that is `CreateJwtAccessToken`. `GetOrCreateJwtAccessTokenAsync` is just the cache manager and in my view should know nothing about what is caching or why. It's not so much about the method calling chain lenght, but about separation of concerns I think.

With your proposal, I'd also worry about code duplication in the creation of the token, everything, but the `aud` and `scope` claims is the same, and I think that would be harder to miss if we needed to add another common claim, we could add it to one of the methods, but not the other. I prefer if we have one single place where we create the token.

Reply:
Ok, thanks for explaining your point, that makes sense.